renderer/backend/InteractionStage.cpp

```diff
@@ -0,0 +1,560 @@
+	};
+}
+
+void InteractionStage::LoadInteractionShader( GLSLProgram *shader, const idStr &baseName, bool bindless ) {
+	idDict defines;
+	defines.Set( "MAX_SHADER_PARAMS", idStr::Fmt( "%d", maxShaderParamsArraySize ) );
+	if (bindless) {
+		defines.Set( "BINDLESS_TEXTURES", "1" );
+	}
+	shader->InitFromFiles( "stages/interaction/" + baseName + ".vs.glsl", "stages/interaction/" + baseName + ".fs.glsl", defines );
+	InteractionUniforms *uniforms = shader->GetUniformGroup<InteractionUniforms>();
+	uniforms->lightProjectionCubemap.Set( TU_LIGHT_PROJECT_CUBE );
+	uniforms->lightProjectionTexture.Set( TU_LIGHT_PROJECT );
+	uniforms->lightFalloffCubemap.Set( TU_LIGHT_FALLOFF_CUBE );
+	uniforms->lightFalloffTexture.Set( TU_LIGHT_FALLOFF );
+	uniforms->ssaoTexture.Set( TU_SSAO );
+	uniforms->stencilTexture.Set( TU_SHADOW_STENCIL );
+	uniforms->depthTexture.Set( TU_SHADOW_DEPTH );
+	uniforms->shadowMap.Set( TU_SHADOW_MAP );
+	if (!bindless) {
+		uniforms->normalTexture.Set( TU_NORMAL );
+		uniforms->diffuseTexture.Set( TU_DIFFUSE );
+		uniforms->specularTexture.Set( TU_SPECULAR );
+	}
+	shader->BindUniformBlockLocation( 0, "ViewParamsBlock" );
+	shader->BindUniformBlockLocation( 1, "PerDrawCallParamsBlock" );
+	shader->BindUniformBlockLocation( 2, "ShadowSamplesBlock" );
+}
+
+
+InteractionStage::InteractionStage( DrawBatchExecutor *drawBatchExecutor )
+	: drawBatchExecutor( drawBatchExecutor ), interactionShader( nullptr )
+{}
+
+void InteractionStage::Init() {
+	maxShaderParamsArraySize = drawBatchExecutor->MaxShaderParamsArraySize<ShaderParams>();
+	
+	ambientInteractionShader = programManager->LoadFromGenerator( "interaction_ambient", 
+		[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.ambient", false ); } );
+	stencilInteractionShader = programManager->LoadFromGenerator( "interaction_stencil", 
+		[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.stencil", false ); } );
+	shadowMapInteractionShader = programManager->LoadFromGenerator( "interaction_shadowmap", 
+		[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.shadowmap", false ); } );
+	if (GLAD_GL_ARB_bindless_texture) {
+		bindlessAmbientInteractionShader = programManager->LoadFromGenerator( "interaction_ambient_bindless", 
+			[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.ambient", true ); } );
+		bindlessStencilInteractionShader = programManager->LoadFromGenerator( "interaction_stencil_bindless", 
+			[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.stencil", true ); } );
+		bindlessShadowMapInteractionShader = programManager->LoadFromGenerator( "interaction_shadowmap_bindless", 
+			[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.shadowmap", true ); } );
+	}
+
+	qglGenBuffers( 1, &poissonSamplesUbo );
+	qglBindBuffer( GL_UNIFORM_BUFFER, poissonSamplesUbo );
+	qglBufferData( GL_UNIFORM_BUFFER, 150 * sizeof(idVec2), nullptr, GL_STATIC_DRAW );
```

Comment:
Yes, good point. I'm not 100% certain, but since the buffer technically gets bound to the shader state even if soft shadows are deactivated, I wanted to play it safe to avoid potential issues that only affect e.g. AMD cards and that I thus wouldn't notice. But you're right in that it certainly doesn't need to be initialized to its max size...